[nix] Flake

jbboehr opened the pull request:
### Todo

- [x] `all` - ~~can't export all the binaries because they collide - could move them to `rosettaboy-<lang>-<build>`. Should they even be in `packages.default`? Maybe add another `packages.all`?~~ later
- [x] `all` - readme update
- [x] `ci` - hashFiles should probably run on flake.nix / ${{ inputs.working-dir}}/derivation.nix
- [x] `nix` - ~~could use a nix formatter. Personally, I've been using [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt) but there are [a few others](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-fmt.html). Currently leaning towards [alejandra](https://github.com/kamadorueda/alejandra) but they both do things I don't care for and are completely unconfigurable by design :shrug:~~ later
- [x] `nix` - shim `shell.nix` and `default.nix` with [flake-compat](https://github.com/edolstra/flake-compat)
- [x] `go` - ~~gomod2nix is using my fork to avoid using an overlay - just use the overlay?~~ - switched to manual/non-flake initialization to avoid introducing an overlay
- [x] `go` - currently failing format
- [x] `utils` - not packaged - could be used to run the entire test suite in `nix flake check`
- [x] `pxd` - not packaged - I gave up for now. devShell still works though.
- [x] `py` - mypyc build isn't working - needs an untagged commit from mypy master. Couldn't get it to work with `overridePythonAttrs`
- [x]  `py` - ~~default package is symlinking all the python code, not just the bin - just use makeWrapper?~~ later
- [x] `py` - currently failing CI

### Notes

* `all` - don't have access to a Mac to test on it :shrug: 
* `nix` - I regretfully deleted your beautiful and comprehensive [comment](https://github.com/shish/rosettaboy/blob/f890797a2d5fa97a0b8371af550b9382c25c4766/shell.nix#L25) in `shell.nix` because it's not quite relevant to flakes - not sure if you want to re-elaborate
* `go` - will need to run `gomod2nix` if the dependencies are changed
* `nim` - doesn't seem to have a `*2nix`, so nimble packages may get out of sync with nix
* `py` - made some changes to `setup.py` and the source files to get it to build with `buildPythonApplication`
* `zig` - I, again, regretfully deleted your comprehensive [comment](https://github.com/shish/rosettaboy/blob/f890797a2d5fa97a0b8371af550b9382c25c4766/zig/shell.nix#L3) in `zig/shell.nix` as it's not quite relevant to flakes - not sure if you want to re-elaborate. SDL2 lowercase doesn't seem to be an issue in CI or locally on NixOS though, ~~not sure what's up with that~~ ~~looks like a Mac thing~~ tried to pull in the previous sdl2-lowercase but untested.
* `utils` - added blargg to flake checks but not currently using `utils/blargg.py` - the interface doesn't quite jive with the Nix directory structure
* `utils` - changed blargg and bench to use roms managed by the nix flake by default, if in a dev shell


#### Default package with `symlinkJoin`:

```bash
$ nix build .
$ find result/bin/ | sort
result/bin/
result/bin/rosettaboy-c
result/bin/rosettaboy-cpp
result/bin/rosettaboy-go
result/bin/rosettaboy-nim
result/bin/rosettaboy-php
result/bin/rosettaboy-py
result/bin/.rosettaboy-py-wrapped
result/bin/rosettaboy-rs
result/bin/rosettaboy-zig
```

#### Closure size

This might only be an issue on my system due to using pipewire but the closure size for the built binary is a bit large right now:

```bash
$ nix path-info  -rsSh '.#c' | tail -3
/nix/store/wfi3pk89p0mzp1gb067by595nqlrpax2-pipewire-0.3.60-lib                    9.0M  539.8M
/nix/store/z4xwhv1zi3maa2dmvghvs2xmi45mhss3-SDL2-2.24.2                            2.3M  542.1M
/nix/store/jxiqfj8m8ybc6m367dmx58w4xw2mg8g4-rosettaboy-c                         283.4K  542.4M
```

~~I might see if static linking SDL2 can keep the final closure size down at some point in the future.~~ Looks like it just needs to link into a lot of system-level dependencies for hardware support [SDL2](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/SDL2/default.nix).

Closes #111

jbboehr commented:
@shish I think I've gotten it to a state where it's almost ready to merge. The only remaining blocker is it needs a touch-up on Mac since I don't have one - hopefully someone can make a pass at it. Everything else I think can be left as-is or addressed later.

cc @rrbutani

jbboehr commented:
These are the current cases of Mac-specific behavior:

```bash
$ rg isDarwin
zig/derivation.nix
45:    ++ lib.optional (!stdenv.isDarwin) SDL2
46:    ++ lib.optional stdenv.isDarwin SDL2-lowercase
50:    ++ lib.optional (!stdenv.isDarwin) autoPatchelfHook

rs/derivation.nix
17:  devTools = [ rustfmt rustc cargo ] ++ lib.optional stdenv.isDarwin [libiconv];

nim/derivation.nix
41:    ++ lib.optional (!stdenvNoCC.isDarwin) sdl2;

c/derivation.nix
30:    ++ lib.optional (!stdenv.isDarwin) autoPatchelfHook;

utils/derivation.nix
30:      ++ lib.optional (!stdenvNoCC.isDarwin) elfutils;

cpp/derivation.nix
22:    ++ lib.optional (!stdenv.isDarwin) autoPatchelfHook;
```

rrbutani commented on .github/workflows/common.yml:
```diff
@@ -111,16 +114,41 @@ jobs:
+      run: nix develop ..#${{ inputs.workdir }} --command bash -c './build.sh && ./rosettaboy-release --help'
+    - name: Flake Checks
+      if: inputs.nix-flake
+      # we need to do this to parallelize the checks
+      run: nix eval --impure --expr 'builtins.currentSystem' | xargs -i nix build -L ".#checks.{}.${{ inputs.workdir }}"
```
https://github.com/rrbutani/rosettaboy/commit/bc5697872e8535662138ca55ded9d9ce29046bbf

rrbutani reviewed (changes requested):
@jbboehr This is looks really good!

I left some suggestions and some notes about fixes I had to apply to get things to build on macOS. The GitHub PR suggestion machinery is a little annoying to use to write/review longer suggested changes so I starting making commits for them on [this branch](https://github.com/rrbutani/rosettaboy/commits/nix-flake) and I've linked the relevant commits where appropriate; feel free to cherry-pick anything you think is useful!

`nix flake check` still isn't _totally_ clean for me on `aarch64-darwin` and `x86_64-darwin` (nim and php have build failures) but this is definitely great progress!

rrbutani commented on cpp/derivation.nix:
```diff
@@ -0,0 +1,38 @@
+
+  src = gitignoreSource ./.;
+
+  enableParallelBuilding = true;
```
Also https://github.com/rrbutani/rosettaboy/commit/406fbd2e42d202d82ef3017429e0d2ff3ae52d19#diff-f31da973b78814f00aaa5be1e5bf48bf525632fe9a9d458b4ae21f6028e9e477L22; I had missed the one in `c`.

rrbutani commented:
Fixed `nim` on macOS: https://github.com/rrbutani/rosettaboy/commit/812198a96f6535940c7998cf65b7e0ce7b7dbf58

Which just leaves PHP (fails during check).

rrbutani commented:
Okay! Couple more fixes:
  - https://github.com/rrbutani/rosettaboy/commit/16c4517a0e935bc7b2a33a6fc779e1b5405a83fc
  - https://github.com/rrbutani/rosettaboy/commit/85603fcb4d9a470395c3cc480de886628542399f

And now `nix flake check` passes on `aarch64-darwin` and `x86_64-darwin`.

jbboehr commented:
@rrbutani Thanks for checking that out! I should get a chance to work on integrating your changes soon.

jbboehr commented on c/derivation.nix:
```diff
@@ -25,8 +25,9 @@ stdenv.mkDerivation {
-    ++ lib.optionals debugSupport devTools;
+  nativeBuildInputs = [ cmake pkg-config ]
+    ++ lib.optionals debugSupport devTools
+    ++ lib.optional (!stdenv.isDarwin) autoPatchelfHook;
```
I'm not sure. I could swear I needed it at one point but it seems to work without it.

jbboehr commented on flake.nix:
```diff
@@ -0,0 +1,209 @@
+    inherit (builtins) mapAttrs;
+    inherit (lib) hiPrio filterAttrs;
+    inherit (gitignore.lib) gitignoreSource;
+    gomod2nix' = rec {
+      gomod2nix = pkgs.callPackage "${gomod2nix-src}" { inherit (lib) buildGoApplication mkGoEnv; };
+      lib = pkgs.callPackage "${gomod2nix-src}/builder" { inherit gomod2nix; };
+    };
```
I was trying to avoid using an overlay if possible.

See: https://github.com/nix-community/poetry2nix/commit/05b788f7b9beaa9014c9d794acf004e23ad571af
See: https://github.com/nix-community/gomod2nix/issues/102

rrbutani replied:
I saw your `gomod2nix` issue; as zimbatm notes in their blog post, I think it makes a lot of sense to be wary of using overlays in flakes that other people are likely to depend on but less so for flakes that are just providing packages to users. I think `rosettaboy` falls pretty comfortably into the latter category; it seems unlikely that many users will pull `rosettaboy` into their own flakes.

I don't feel strongly about this though; whatever approach you prefer is fine with me.

jbboehr commented:
@rrbutani I've integrated a fair number of your changes but ran into some issues with zig and mypyc. I should get a chance to try again soon.

rrbutani commented on flake.nix:
```diff
@@ -0,0 +1,209 @@
+      
+      zig-fast = mkZig { fastSupport = true; };
+      zig-safe = mkZig { safeSupport = true; };
+      zig = hiPrio zig-fast;
```
@jbboehr I do think the changes above that aren't overlay related are still worthwhile; I can split https://github.com/rrbutani/rosettaboy/commit/e8711d9da1671a3d8fc6cfce9f81194c8fa2212b so that the overlay changes are separate from the `newScope` changes.

rrbutani commented on py/derivation.nix:
```diff
@@ -0,0 +1,33 @@
+  nativeBuildInputs = lib.optional mypycSupport pythonPackages.mypy;
+
+  passthru.python = python.withPackages (p: runtimeDeps ++ devDeps);
+  passthru.devTools = [ python ];
```
Just a heads up: earlier `passthru.devtools` used `python.withPackages ...` while this just uses the unmodified python interpreter.

https://github.com/shish/rosettaboy/blob/2f7b52686962b453a438e956cbfc4429374c80f1/py/derivation.nix#L15-L25

---

I don't think this actually makes a difference since `devTools` is only used for the devShell and `propagatedBuildInputs` gets [picked up anyways by `mkShell`](https://github.com/NixOS/nixpkgs/blob/f3b7d6414b06eb493ed7f258a6938be440b7cdaf/pkgs/build-support/mkshell/default.nix#L38) (and even if the "bare" interpreter gets put on `$PATH` ahead of the one including the packages it's still fine since the interpreter's [setup hook](https://github.com/NixOS/nixpkgs/blob/6979a72840a1f6f034cb6a11a3d114b17f792ad4/pkgs/development/interpreters/python/setup-hook.sh#L15-L17) will make sure they're added to `$PYTHONPATH`; it's only in cases where `$PYTHONPATH` is cleared or not passed to child processes that invoke `python` that there's an observable difference afaik).

---

I think it's probably fine to drop `devTools` from this derivation.

jbboehr replied:
Fails in CI after I removed the override in flake's devShells: https://github.com/shish/rosettaboy/actions/runs/4140614710/jobs/7159436299 :shrug: Just going to leave it for now.

rrbutani commented on cpp/derivation.nix:
```diff
@@ -0,0 +1,36 @@
+  src = gitignoreSource ./.;
+
+  buildInputs = [ SDL2 ];
+  nativeBuildInputs = [ cmake pkg-config ]
+    ++ lib.optional (!stdenv.isDarwin) autoPatchelfHook;
```
Whoops, forgot to remove this one; `autoPatchelfHook` doesn't appear to be needed here either (tested on `aarch64-linux`):

```suggestion
  nativeBuildInputs = [ cmake pkg-config ];
```

rrbutani commented on cpp/derivation.nix:
```diff
@@ -0,0 +1,36 @@
+, stdenv
+, cmake
+, SDL2
+, autoPatchelfHook
```
(see below)

```suggestion
```

rrbutani reviewed (commented):
Couple of nits; nothing major. I think only these commits remain unmerged:
  - [ ] `mypyc` fixes: https://github.com/rrbutani/rosettaboy/commit/199e73deb1749c82d530fa0d37cd8582f5cbc149
  - [ ] update the zig toolchain to pick up `NIX_CFLAGS_COMPILE` fixes for macOS: https://github.com/rrbutani/rosettaboy/commit/b74e9575b1af903a1f2d99718df67eee3b893c8e
  - [ ] zig `derivation.nix` fixes for macOS: https://github.com/rrbutani/rosettaboy/commit/13663158d23b8317bd09e3bc2d53072c64259337
  - [ ] drop the zig SDL2 fix: https://github.com/rrbutani/rosettaboy/commit/0d3709ec438a962db678ebe95bde25da99d4a040
  - [ ] https://github.com/rrbutani/rosettaboy/commit/16c4517a0e935bc7b2a33a6fc779e1b5405a83fc (optional)

---

Thanks for pushing on this @jbboehr!

jbboehr commented:
@rrbutani I managed to get your zig changes working, I'll pull them into the PR soon.

I pulled in your mypyc changes [[branch]](https://github.com/jbboehr/rosettaboy/tree/py/mypyc-again) but it seems like it's not using mypyc:

```bash
$ nix run .#py-mypyc -- --headless --frames 100 $GB_DEFAULT_BENCH_ROM
Emulated   100 frames in 11.85s (8fps)

$ nix run .#py -- --headless --frames 100 $GB_DEFAULT_BENCH_ROM
Emulated   100 frames in 12.34s (8fps)
```

mypy takes about 15 minutes to compile on my machine. Might be better to wait until it's in nixpkgs.

It might be my fault as I had tried remapping the package name in `setup.py` earlier. I'm just not a python guy and haven't figured out how to remap it for mypycify. Or maybe just revert the package name mapping.

```bash
$ nix build -L .#py-mypyc
$ find result/lib/python3.11/site-packages
result/lib/python3.11/site-packages
result/lib/python3.11/site-packages/14f66929a8af613578ab__mypyc.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info/METADATA
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info/RECORD
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info/INSTALLER
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info/direct_url.json
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info/top_level.txt
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info/REQUESTED
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info/WHEEL
result/lib/python3.11/site-packages/rosettaboy-0.0.0.dist-info/entry_points.txt
result/lib/python3.11/site-packages/src
result/lib/python3.11/site-packages/src/ram.cpython-311-x86_64-linux-gnu.so    <- named src here
result/lib/python3.11/site-packages/src/main.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/consts.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/gameboy.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/clock.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/cpu.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/buttons.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/args.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/__init__.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/gpu.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/cart.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/src/errors.cpython-311-x86_64-linux-gnu.so
result/lib/python3.11/site-packages/rosettaboy
result/lib/python3.11/site-packages/rosettaboy/buttons.py     <- named rosettaboy here
result/lib/python3.11/site-packages/rosettaboy/errors.py
result/lib/python3.11/site-packages/rosettaboy/consts.py
result/lib/python3.11/site-packages/rosettaboy/gpu.py
result/lib/python3.11/site-packages/rosettaboy/gameboy.py
result/lib/python3.11/site-packages/rosettaboy/args.py
result/lib/python3.11/site-packages/rosettaboy/cart.py
result/lib/python3.11/site-packages/rosettaboy/clock.py
result/lib/python3.11/site-packages/rosettaboy/main.py
result/lib/python3.11/site-packages/rosettaboy/cpu.py
result/lib/python3.11/site-packages/rosettaboy/__pycache__
result/lib/python3.11/site-packages/rosettaboy/__pycache__/errors.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/ram.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/consts.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/cpu.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/gpu.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/cart.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/clock.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/main.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/__init__.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/buttons.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/gameboy.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/__pycache__/args.cpython-311.pyc
result/lib/python3.11/site-packages/rosettaboy/ram.py
result/lib/python3.11/site-packages/rosettaboy/__init__.py
```

jbboehr commented:
@rrbutani This was the issue I was having earlier with zig, seems to be intermittent because the CI just passed earlier with no changes. AFAICT started happening since the toolchain was updated but hard to tell. Might be better to revert to an older release. Maybe it's cache-related, I don't know.

See: https://github.com/shish/rosettaboy/actions/runs/4141038600/jobs/7160260155

```
Run nix run .#zig -- --help
/home/runner/work/_temp/0844677d-b414-4746-aff7-294198b4b375.sh: line 1:  4585 Illegal instruction     (core dumped) nix run .#zig -- --help
```

jbboehr commented:
I think I got `pxd` packaged but there's something weird going on - exit code is ok though. Could be my fault.

```bash
$ /nix/store/wpkn4j3jv25057sncr248ppvdr0irfdg-rosettaboy-pxd/bin/rosettaboy-pxd --turbo --silent --headless --frames 200 /nix/store/gmm7glcs2qsfpvdvrnkw6ci1hqkidf11-source/blargg-cpu-instructions/04-op_r,imm.gb && echo ok
04-op r,imm


Passed
Traceback (most recent call last):
  File "src/cpu.py", line 468, in src.cpu.CPU.tick_instructions
  File "src/cpu.py", line 1068, in src.cpu.CPU.tick_main
src.errors.UnitTestPassed: Unit test passed
Exception ignored in: 'src.cpu.CPU.tick'
Traceback (most recent call last):
  File "src/cpu.py", line 468, in src.cpu.CPU.tick_instructions
  File "src/cpu.py", line 1068, in src.cpu.CPU.tick_main
src.errors.UnitTestPassed: Unit test passed
Traceback (most recent call last):
  File "src/cpu.py", line 468, in src.cpu.CPU.tick_instructions
  File "src/cpu.py", line 1070, in src.cpu.CPU.tick_main
src.errors.UnitTestFailed: Unit test failed
Exception ignored in: 'src.cpu.CPU.tick'
Traceback (most recent call last):
  File "src/cpu.py", line 468, in src.cpu.CPU.tick_instructions
  File "src/cpu.py", line 1070, in src.cpu.CPU.tick_main
src.errors.UnitTestFailed: Unit test failed
Emulated   200 frames in  0.78s (256fps)
ok
```

jbboehr commented:
@shish @rrbutani I think this is ready to go barring another check on Mac.

Some other things that could be done / general notes

- `nix` - build docker images using Nix. I looked into it briefly but they're ~700MB on my machine, probably due to SDL pulling in lots of system level dependencies I noted above about the closure size (like pipewire, mesa, etc). [prototype](https://github.com/jbboehr/rosettaboy/commit/bc4126ba1d295f697dfec0d0bcc839682dee5ad9)
- `nix` - pre-commit hooks for the languages. It's a little tricky as `mkShell` merges the shell hooks.
- `nix` - use makeWrapper, `ln`, or something like it instead of symlinkJoin to build the default package as only the binaries are required. Doesn't really matter a whole lot but would be a bit cleaner.
- `nix` - decide whether to export the non-default builds (e.g. `cpp-lto`) in `packages.default`
- `nix` - pick a formatter. I'm on the fence. Samples: [alejandra](https://github.com/jbboehr/rosettaboy/commit/84f6289a0b0e14103d40e151d73fee22373cabb7) [nixpkgs-fmt](https://github.com/jbboehr/rosettaboy/commit/43735cdeb9891b8ea2f189e6180b5941410c7997).
- `nix` - use `utils/blargg.py` to run blargg in `nix flake check`. It needs some changes to work with the nix directory structure and I'm just not a Python guy.
- `nix` - expose a package and/or check to run `utils/bench.py`. It also needs some changes to work with the nix directory structure.
- `pxd` - figure out what's up with the test output mentioned in https://github.com/shish/rosettaboy/pull/119#issuecomment-1426916617
- `py` - mypy takes about 15 minutes to compile on my machine and in CI. Might be better to disable the mypyc nix build until mypy is in nixpkgs (or set up cachix?). Could also disable compiling mypy with mypyc maybe [see](https://github.com/NixOS/nixpkgs/commit/ad26cb9ee8842a69884212d098e6f079c91c0f96#diff-390cc6eb464de84abeb476ddc369403ee736b8d1d82c481b5aace4be980aec88R56)
- `zig` - the illegal instruction issue has not returned :confetti_ball: 

but I don't think any of these are blockers.

will-ca commented:
To be honest, though I contributed the original Nix shells, as an outside observer at this point— Not to dismiss what Nix Flakes *can* do, nor that it's cool that they can do that.— This feels like a lot of work and increased maintenance and barrier of entry for marginal practical gain.

E.G. Just speaking on my personal experience, I suspect if I had seen this at the time, it may have intimidated me enough to deter me from contributing the Cythonized implementation.

shish commented:
Complexity of the nix bits is certainly something to think about; but also this PR seems to have a bunch of stuff that doesn't all need to happen in one go, which makes it harder to understand what's being changed -- I've cherry-picked most of the independent parts (updates to utils scripts, allowing github actions to specify nim or zig versions, rebuilding everything when common.yml changes, etc etc); once that's done I'll have a look at cherry-picking the "use zig 0.11" bits, which should also be independent but slightly trickier to cherry-pick; then hopefully what's left after that is just the flakes by themselves, and fingers crossed the flakes by themselves are less intimidating (I confess that I too don't really understand them, hence getting everything-except-the-flakes merged first :) )

jbboehr commented:
@will-ca Virtually all the complexity is in building the packages in Nix, the Nix way (TM). @rrbutani's original 3 commits were sufficient for just the devShells. Personally, I see value in declaring build instructions in a reproducible manner, which is why I switched to NixOS ~6 years ago.

I was initially hesitant to upgrade to flakes, myself, them still being "experimental", but made the switch a few months ago and have been very pleased. No more [of this nonsense](https://github.com/jbboehr/php-handlebars/blob/c270cba54ec0425865930e4e414e63ae37e83027/default.nix#L8-L49).

In my personal projects, I don't expect drive-by contributors to work on the Nix parts, but this is @shish's project so he'll have to make the call on whether or not he wants to do that.

Could also mark the Nix builds with [continue-on-error](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error), I suppose.

shish commented:
> Could also mark the Nix builds with continue-on-error, I suppose

Thinking along similar lines, I wonder if it's possible to skip the nix parts completely if `<language>/flake.nix` doesn't exist 🤔  (Adding a boolean config option in common.yml works, but being totally automatic would be nicer)

My current high-level thinking is that I like all the nice things that come with nix and flakes, and I'm gradually starting to use them for all my personal development-environment-setup tasks, but I wouldn't want nix expertise to be _required_ at any point -- ideally the only barrier to entry for contributing a new implementation should be familiarity with that language + enough git knowledge to create a pull request

will-ca commented:
> Personally, I see value in declaring build instructions in a reproducible manner, which is why I switched to NixOS ~6 years ago.

> but I wouldn't want nix expertise to be _required_ at any point -- ideally the only barrier to entry for contributing a new implementation should be familiarity with that language + enough git knowledge to create a pull request

I think I see the value in theory, but in practice I'm not sure it's worth the cost of making the build process much harder to understand/more intimidating for anyone without that relatively specialized knowledge of "The Nix Way".

If Nix Flakes were the standard workflow, then presumably we would be horrified at the thought of doing things the non-reproducible way. But as it is, the non-reproducible workflow is the standard, so any deviation from that, even if strictly better on technical merits, also carries a logistical cost in relation to the rest of the ecosystem. I can only read so many API docs in a day, and while a couple hundred lines of Nix is certainly better than fighting with disparate distros and build systems, at some point I feel it would start to eat into the resources for writing application code and deploying on non-Nix systems. …Nix sans Flakes is both innovative and a large practical improvement over the old way; Nix Flakes may be just as innovative but it feels like less of a relative practical improvement compared to Nix sans Flakes (mainly because Nix sans Flakes is already quite good in most cases).

'Just my two cents, and my confused attempt to understand the landscape. Innovation vs. convention, etc.

jbboehr commented:
In my latest commit I:

* Squashed the entire branch - it was getting out of hand and most of the non-Nix commits have been merged into master already
* Brought back the old `shell.nix` for each language (except c) :shrug:
* Split Nix into two (three technically) workflows which check for the presence of a shell.nix or derivation.nix and run the shell-only checks or flake-only checks respectively.
* Added support for `[skip format]` `[skip build]` `[skip nix]` `[skip nix shell]` `[skip nix flake]` to the workflows - feel free to remove of course

shish commented:
This all seems to work for me, and my brain-buffer is full up from trying to keep track of wide-spread changes - so I'm merging now, and we can fix forward with smaller PRs for future work /o/
